Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -1263,4 +1275,237 @@ lemma nonsingularLift_addMap {P Q : PointClass F} (hP : W.NonsingularLift P)
 
 end Addition
 
+section Point
+
+/-! ### Nonsingular rational points -/
+
+variable (W') in
+/-- A nonsingular rational point on `W'`. -/
+@[ext]
+structure Point where
+  {point : PointClass R}
+  (nonsingular : W'.NonsingularLift point)
+
+/-- The point class underlying a nonsingular rational point on `W'`. -/
+add_decl_doc Point.point
+
+/-- The nonsingular condition underlying a nonsingular rational point on `W'`. -/
+add_decl_doc Point.nonsingular
+
+namespace Point
+
+lemma mk_point {P : PointClass R} (h : W'.NonsingularLift P) : (mk h).point = P :=
+  rfl
+
+instance instZeroPoint [Nontrivial R] : Zero W'.Point :=
+  ⟨⟨nonsingularLift_zero⟩⟩
+
+lemma zero_def [Nontrivial R] : (0 : W'.Point) = ⟨nonsingularLift_zero⟩ :=
+  rfl
+
+lemma zero_point [Nontrivial R] : (0 : W'.Point).point = ⟦![1, 1, 0]⟧ :=
+  rfl
+
+/-- The map from a nonsingular rational point on a Weierstrass curve `W'` in affine coordinates
+to the corresponding nonsingular rational point on `W'` in Jacobian coordinates. -/
+def fromAffine [Nontrivial R] : W'.toAffine.Point → W'.Point
+  | 0 => 0
+  | Affine.Point.some h => ⟨(nonsingularLift_some ..).mpr h⟩
+
+lemma fromAffine_zero [Nontrivial R] : fromAffine 0 = (0 : W'.Point) :=
+  rfl
+
+lemma fromAffine_some [Nontrivial R] {X Y : R} (h : W'.toAffine.Nonsingular X Y) :
+    fromAffine (Affine.Point.some h) = ⟨(nonsingularLift_some ..).mpr h⟩ :=
+  rfl
+
+/-- The negation of a nonsingular rational point on `W`.
+Given a nonsingular rational point `P` on `W`, use `-P` instead of `neg P`. -/
+def neg (P : W.Point) : W.Point :=
+  ⟨nonsingularLift_negMap P.nonsingular⟩
+
+instance instNegPoint : Neg W.Point :=
+  ⟨neg⟩
+
+lemma neg_def (P : W.Point) : -P = P.neg :=
+  rfl
+
+lemma neg_point (P : W.Point) : (-P).point = W.negMap P.point :=
+  rfl
+
+/-- The addition of two nonsingular rational points on `W`.
+Given two nonsingular rational points `P` and `Q` on `W`, use `P + Q` instead of `add P Q`. -/
+noncomputable def add (P Q : W.Point) : W.Point :=
+  ⟨nonsingularLift_addMap P.nonsingular Q.nonsingular⟩
+
+noncomputable instance instAddPoint : Add W.Point :=
+  ⟨add⟩
+
+lemma add_def (P Q : W.Point) : P + Q = P.add Q :=
+  rfl
+
+lemma add_point (P Q : W.Point) : (P + Q).point = W.addMap P.point Q.point :=
+  rfl
+
+end Point
+
+end Point
+
+section Affine
+
+/-! ### Equivalence with affine coordinates -/
+
+open scoped Classical
+
+namespace Point
+
+variable (W) in
+/-- The map from a point representative that is nonsingular on a Weierstrass curve `W` in Jacobian
+coordinates to the corresponding nonsingular rational point on `W` in affine coordinates. -/
+noncomputable def toAffine (P : Fin 3 → F) : W.toAffine.Point :=
+  if hP : W.Nonsingular P then if hPz : P z = 0 then 0 else
+    Affine.Point.some <| (nonsingular_of_Z_ne_zero hPz).mp hP else 0
```

Comment:
How about
```suggestion
  if hP : W.Nonsingular P ∧ P z ≠ 0 then
    .some <| (nonsingular_of_Z_ne_zero hP.2).mp hP.1 else 0
```
? Or maybe you want `split_ifs` to always give you three cases automatically ...

Reply:
I think this is a great idea, since one of these is a junk value and I rarely do `split_ifs`.

Originally `toAffine` took in a *proof* `hP : W.Nonsingular P` and returns the element `if hPz : P z = 0 then 0 else .some <| (nonsingular_of_Z_ne_zero hPz).mp hP`. This turned out to be impossible to work with, since `toAffine_of_equiv` would have to take in a `HEq` proof that `hP` and `hQ` are the same, and I got stuck many times... This makes the infoview a lot more friendly too: we'd be seeing `toAffine W P` rather than `toAffine ...`. In a future PR I might decide to do something similar for other definitions that take in proofs e.g. `Affine.Point.some` and give them junk values when needed.

Note that I have a local branch that combines the conditions for `Affine.slope` similar to this, but haven't made a PR.